build_tools/cmake/build_and_test_asan.sh

```diff
@@ -25,110 +24,131 @@ IREE_ENABLE_ASSERTIONS="${IREE_ENABLE_ASSERTIONS:-ON}"
-export IREE_VULKAN_F16_DISABLE=${IREE_VULKAN_F16_DISABLE:-1}
-# Respect the user setting, but default to skipping tests that require Nvidia GPU.
-export IREE_NVIDIA_GPU_TESTS_DISABLE=${IREE_NVIDIA_GPU_TESTS_DISABLE:-1}
-
-# Tests to exclude by label. In addition to any custom labels (which are carried
-# over from Bazel tags), every test should be labeled with its directory.
-declare -a label_exclude_args=(
-  # Exclude specific labels.
-  # Put the whole label with anchors for exact matches.
-  # For example:
-  #   ^nokokoro$
-  ^nokokoro$
-
-  # Exclude all tests in a directory.
-  # Put the whole directory with anchors for exact matches.
-  # For example:
-  #   ^bindings/python/iree/runtime$
-
-  # Exclude all tests in some subdirectories.
-  # Put the whole parent directory with only a starting anchor.
-  # Use a trailing slash to avoid prefix collisions.
-  # For example:
-  #   ^bindings/
-)
-
-# IREE_VULKAN_DISABLE is handled separately as we run Vulkan and non-Vulkan
-# tests in separate ctest commands anyway.
-if [[ "${IREE_CUDA_DISABLE?}" == 1 ]]; then
-  label_exclude_args+=("^driver=cuda$")
-fi
-if [[ "${IREE_VULKAN_F16_DISABLE?}" == 1 ]]; then
-  label_exclude_args+=("^vulkan_uses_vk_khr_shader_float16_int8$")
-fi
-if [[ "${IREE_NVIDIA_GPU_TESTS_DISABLE}" == 1 ]]; then
-  label_exclude_args+=("^requires-gpu-nvidia$")
-fi
-
-label_exclude_regex="($(IFS="|" ; echo "${label_exclude_args[*]?}"))"
-
-vulkan_label_regex='^driver=vulkan$'
-
-cd ${BUILD_DIR?}
-
-echo "*** Running main project ctests that do not use the Vulkan driver *******"
-ctest \
-  --timeout 900 \
-  --output-on-failure \
-  --no-tests=error \
-  --label-exclude "${label_exclude_regex}|${vulkan_label_regex}"
-
-echo "******************** llvm-external-projects tests ***********************"
-cmake --build . --target check-iree-dialects -- -k 0
-
-if [[ "${IREE_VULKAN_DISABLE?}" == 0 ]]; then
-  echo "*** Running ctests that use the Vulkan driver, with LSAN disabled *****"
-  # Disable LeakSanitizer (LSAN) because of a history of issues with Swiftshader
-  # (#5716, #8489, #11203).
-  ASAN_OPTIONS=detect_leaks=0 \
-    ctest \
-      --timeout 900 \
-      --output-on-failure \
-      --no-tests=error \
-      --label-regex "${vulkan_label_regex}" \
-      --label-exclude "${label_exclude_regex}"
-fi
+# We run everything twice, for each boolean value of `asan_in_bytecode_modules`:
+#  - When asan_in_bytecode_modules=false, ASAN is only enabled in the C/C++ code
+#    (compiler and runtime), not in bytecode modules. The default embedded-ELF
+#    path is used.
+#  - When asan_in_bytecode_modules=true, ASAN is also enabled in bytecode
+#    modules. The system-ELF path is used (required for ASAN-in-modules).
+for asan_in_bytecode_modules in false true; do
+  if [[ $asan_in_bytecode_modules == true ]]; then
+    asan_in_bytecode_modules_ON_OFF=ON
+  else
+    asan_in_bytecode_modules_ON_OFF=OFF
+  fi
```

Comment:
Why have two variables here?

```suggestion
for asan_in_bytecode_modules_ON_OFF in ON OFF; do
```

Or

```suggestion
for IREE_BYTECODE_MODULE_ENABLE_ASAN in ON OFF; do
```

would be more like our other variables (although I see why you're trying to be explicit about the specific values this could take)

Reply:
Sorry, that was a remnant from an earlier state.